lib/handler/mruby/channel.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2017 Ritta Narita
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include <mruby.h>
+#include <mruby/array.h>
+#include <mruby/error.h>
+#include <mruby/hash.h>
+#include <mruby/string.h>
+#include <mruby/class.h>
+#include <mruby/variable.h>
+#include "h2o/mruby_.h"
+#include "embedded.c.h"
+
+struct st_h2o_mruby_channel_context_t {
+    h2o_mruby_context_t *ctx;
+    mrb_value receiver;
+};
+
+static void attach_receiver(struct st_h2o_mruby_channel_context_t *ctx, mrb_value receiver)
+{
+    assert(mrb_nil_p(ctx->receiver));
```

Comment:
Could you please elaborate how we ensure that a Channel will have at most one listeners?

I couldn't find code that provides such guarantee, and wonder what happens if two fibers call `shift` of one channel object.

Reply:
PS. Please note that I am not suggesting that we should support multiple readers. Supporting only one reader is fine, but an error in mruby script should be reported as an mruby exception; not an abrupt termination of the server process.

Reply:
I fixed to raise error when shift is called though it's already registered.
https://github.com/h2o/h2o/pull/1336/commits/2cf40c2ee4c2c78dcef25c245ee9ad1a2004c9c9

Reply:
@naritta Would you mind supporting multiple listener (shifter)? Discussed with @kazuho today, it may be useful in minor cases and can be easily implemented .

To implement that, simply do:
- change receiver to receivers (Array)
- use mrb_ary_push instead of attach_receiver
- use mrb_ary_shift instead of detach_receiver

How do you feel?

Reply:
Sounds cool. I fixed to support multiple listener like above.
https://github.com/h2o/h2o/pull/1336/commits/916cab15f8f88cfb608b769379f6af4139fb8612